Simulate veri-testharness with RISC-V proxy kernel

isa084 opened the pull request:
This pull request introduces support for RISC-V proxy kernel into veri-testharness simulation target. 
It is a follow up to the discussion in https://github.com/openhwgroup/cva6/issues/2426

Highlights : 
1. Successfully tested with cv64a6_imafdc_sv39, however it currently fails with cv32a65x (test programs have been provided for both cases).
2. Changes are localized, non-invasive and transparent. Users who don't need this feature will see negligible build and run-time overhead. 
3. Provides backbone for strengthening proxy kernel support in the future. 

Key changes : 
1. Added a new simulation target veri-testharness-pk : A simplified version of veri-testharnes, with less simulation params.

2. Modifications to cva6.py : 
2.a. Installation of RISC-V proxy kernel : cva6.py now has an option to install RISC-V proxy kernel. The pk installation configuration (e.g. march, mabi) are matched to the hardware configs parsed by cva6.py. This is an attempt to maintain consistency between the proxy kernel and the simulated architecture.
2.b. GCC compile options : Certain header/assembly/C files and compile-time options are skipped when compiling a test program for veri-testharness-pk.

3. Install-pk script : This script is called from cva6.py to to install the proxy kernel. It is lifted off of install-verilator.sh. The installation logs are stored in verif/sim/pk_install.log. 

Test programs : 
Two tests are provided : 
1. verif/sim/hello-cv64a6_imafdc_sv39-verilator.sh (a 64-bit target) : Successfully tested. This simulates hello_world.c on veri-testharness-pk. The logfile shows "Hello World", mirroring the printf statement in hello_world.c
2. verif/sim/hello-cv32a65x-verilator.sh (a 32-bit target) : Failing. This follows the same steps as the 64-bit version, however it encounters a simulation error :  "../machine/mtrap.c:22: machine mode: unhandlable trap 2 @ 0x80001c0e". This needs further debugging and is being shared for bug reproducibility.

valentinThomazic commented:
Hello @isa084, as you know I won't really be availablelater so I am providing as much feedback as I can now, I hope it is not too overwhelming.

1. The test scripts need to be in `verif/regress` instead of `verif/sim` (all test scripts are there)
2. It would be nice to add the mention of "pk" in the test script name
3. It is probably better to remove the cv32a65x test script. It does not work which is fine but I think it is best not to merge non functional things since people will probably expect them to work.
4. Please change the install script for the pk so it does not install the pk in the toolchain directory : for a lot of users they will not have write access to their toolchain directory and it will not work for them (It dit not work for me and I had to change the Makefile). Instead of running `make install` you could just copy the built pk in `tools/pk`
5. Please change the ``verif/sim/Makefile` accordingly (change the path of the pk bin).
6. It is not ideal to install the pk via `verif/sim/cva6.py`. I understand why you did it (you needed the ISA string to compile the pk) but its purpose is to run the simulations not to install the tools I understand you made it so that it will compile for all targets. I have a few ideas of how it could be done another way but this would require more work. Maybe a good first step would be to run the install script in the test script for the 64bits config.
7. On the install pk script, it would probably be better to pick a commit hash (which can be the current top of the master branch) instead of picking the last one. This will ensure that the behaviour is consistent and will not be affected by an update on the riscv-pk repo.
8. We will add a CI job to test the feature, please apply this patch on `.gitlab-ci.yml` and replace the <new_name_of_the_script> part:
```
diff --git a/.gitlab-ci.yml b/.gitlab-ci.yml
index b94585e7..0622f24e 100644
--- a/.gitlab-ci.yml
+++ b/.gitlab-ci.yml
@@ -208,6 +208,19 @@ smoke-hwconfig:
     - source verif/regress/hwconfig_tests.sh
     - python3 .gitlab-ci/scripts/report_pass.py

+hello-pk:
+extends:
+  - .fe_smoke_test
+variables:
+  DASHBOARD_JOB_TITLE: "Hello world with riscv-pk on cv64a6_imafdc_sv39"
+  DASHBOARD_JOB_DESCRIPTION: "Short tests to challenge target configurations"
+  DASHBOARD_SORT_INDEX: 1
+  DASHBOARD_JOB_CATEGORY: "Basic"
+  COLLECT_SIMU_LOGS: 1
+script:
+  - bash verif/regress/<new_name_of_the_script>.sh
+  - !reference [.simu_after_script]
+
 .synthesis_test:
   stage: heavy tests
   timeout: 2 hours
```

isa084 commented:
Thanks a lot for the detailed feedback ! All of the eight points are quite clear. I will incorporate them in my next update.

isa084 commented:
This pull request now provides a working setup for proxy kernel, incorporating all of the following feedbacks from @valentinThomazic :

1. Test scripts moved to `verif/regress`
2. Mentioned `pk` in the test script .
3. Removed non-functional test script for cv32a65x
4. Updated install-pk script so that it installs pk in `tools/pk` instead of RISCV toolchain directory. This was fixed by using `--prefix=$PK_INSTALL_DIR` instead of `--prefix=$RISCV` in install-pk.sh.
5. Updated `verif/sim/Makefile` to be consistent with (4)
6. Removed pk installation from `verif/sim/cva6.py`. Moved installation to test script in (1).
7. For installation of riscv-pk from source, install-pk.sh refers to a recent commit hash on riscv-pk's github repository (dated February 5 2025). 
8. Added a hello world test in `.gitlab-ci.yml`

valentinThomazic commented on verif/sim/cva6.py:
```diff
@@ -358,7 +358,6 @@ def elf2bin(elf, binary, debug_cmd):
   cmd = ("%s -O binary %s %s" % (get_env_var("RISCV_OBJCOPY", debug_cmd = debug_cmd), elf, binary))
   run_cmd_output(cmd.split(), debug_cmd = debug_cmd)
 
-
```
can you add the line back ?

valentinThomazic commented on verif/sim/cva6.py:
```diff
@@ -964,7 +968,6 @@ def load_config(args, cwd):
 
   args.spike_params = get_full_spike_param_args(args.spike_params) if args.spike_params else ""
 
-
```
can you add the line back ?

valentinThomazic commented on verif/regress/hello-verilator-pk-cv64a6_imafdc_sv39.sh:
```diff
@@ -0,0 +1,42 @@
+if [[ "$DV_SIMULATORS" == *"veri-testharness-pk"* ]]; then
+
+  echo "[ riscv-pk ] veri-testharness-pk simulation detected. Installing RISC-V proxy kernel..."
+#  source ./verif/regress/install-pk.sh ${PK_ARCH} ${PK_MABI} 2>&1 | tee ./verif/sim/pk-install.log
```
```suggestion
```

valentinThomazic commented on verif/regress/hello-verilator-pk-cv64a6_imafdc_sv39.sh:
```diff
@@ -0,0 +1,42 @@
+
+TIMEOUT_WALLCLOCK="600"
+TIMEOUT_TICKS="2100000"
+#export TRACE_COMPACT=1
```
```suggestion
```

valentinThomazic commented on verif/regress/hello-verilator-pk-cv64a6_imafdc_sv39.sh:
```diff
@@ -0,0 +1,42 @@
+#  source ./verif/regress/install-pk.sh ${PK_ARCH} ${PK_MABI} 2>&1 | tee ./verif/sim/pk-install.log
+  source ./verif/regress/install-pk.sh ${PK_ARCH} ${PK_MABI} > ./verif/sim/pk-install.log 2>&1
+  echo "[ riscv-pk ] RISC-V proxy kernel installation logs at $(pwd)/verif/sim/pk-install.log"
+#  echo "PK_INSTALL_DIR is ${PK_INSTALL_DIR}"
```
please uncomment or remove

valentinThomazic commented on verif/regress/hello-verilator-pk-cv64a6_imafdc_sv39.sh:
```diff
@@ -0,0 +1,42 @@
+source ./verif/sim/setup-env.sh
+cd ./verif/sim
+
+# CC_OPTS="-static -mcmodel=medany -fvisibility=hidden -nostdlib -nostartfiles -g ../tests/custom/common/syscalls.c ../tests/custom/common/crt.S -I../tests/custom/env -I../tests/custom/common -lgcc"
```
```suggestion
```

valentinThomazic commented on verif/regress/install-pk.sh:
```diff
@@ -0,0 +1,85 @@
+    echo "pk will be installed in '$PK_INSTALL_DIR'"
+    echo "PK_REPO=$PK_REPO"
+    echo "PK_BRANCH=$PK_BRANCH"
+#    echo "VERILATOR_HASH=$VERILATOR_HASH"
```
please uncomment or remove

valentinThomazic commented on verif/regress/install-pk.sh:
```diff
@@ -0,0 +1,85 @@
+    echo "PK_REPO=$PK_REPO"
+    echo "PK_BRANCH=$PK_BRANCH"
+#    echo "VERILATOR_HASH=$VERILATOR_HASH"
+#    echo "VERILATOR_PATCH='$VERILATOR_PATCH'"
```
please uncomment or remove

valentinThomazic commented:
Hello @isa084 
With your changes it worked out of the box, which is great!
I made some suggestions above.
I made a mistake in the patch I gave you and thus the `.gitlab-ci.yml` file is invalid.
Spaces should be inserted at the beginning of each line under the hello_pk job.
Once this is done everything should be good!

valentinThomazic commented:
Hello @isa084, I apologize for the delay.
The hello pk test is not passing because the script name in the CI job does not seem to be correct. Here is the log:

```
$ bash verif/regress/hello-veri-testharness-pk-cv64a6_imafdc_sv39.sh
bash: verif/regress/hello-veri-testharness-pk-cv64a6_imafdc_sv39.sh: No such file or directory
```
It should be `verif/regress/hello-verilator-pk-cv64a6_imafdc_sv39.sh` instead.

valentinThomazic commented:
hello @isa084, there was an issue with the dashboard not showing recent pipeline reports. It is now solved.
There is indeed an issue in the ci script because it is supposed to be comparing spike execution with verilator.

This should do the trick:
```
  script:
    - 
    - bash verif/regress/hello-verilator-pk-cv64a6_imafdc_sv39.sh
    - python3 .gitlab-ci/scripts/report_pass.py
```

valentinThomazic commented on verif/regress/install-pk.sh:
```diff
@@ -0,0 +1,83 @@
+    NUM_JOBS=1
+fi
+
+#PK_PATCH="$ROOT_PROJECT/verif/regress/pk.patch"
```
Is this comment still useful?

valentinThomazic commented:
Hey @isa084, appart from the comment I left, this looks good to me!
I will take some time to talk with @JeanRochCoulon and we will let you know!
Maybe one last good thing would be to remove the "Initial commit" in the PR title

isa084 commented:
@valentinThomazic , I have updated the scripts . Instead of having a script to simulate only for `cv64a6_imafdc_sv39`, I now have a new script `veri-testharness-pk-tests.sh`. This runs two simulations, one after the other - one for `cv64aa6_imafdc_sv39` , and one for `cv32a6_imac_sv32` .

I have also updated the gitlab CI yml file to account for this.

I have also updated `tutorials/running_sim.md` with a section for this simulation mode. Please take a look.

I also added solderpad licenses at the top of the new scripts (i.e. `veri-testharness-pk-tests.sh` and `install-pk.sh`).

This looks complete to me. If you have any other suggestions, let me know. Otherwise, you may merge this to the mainstream.

valentinThomazic commented:
@isa084 this looks good to me, the script output is nice btw.
@JeanRochCoulon LGTM

JeanRochCoulon commented:
Thanks @isa084 and @valentinThomazic for your effort.
LGTM, I will merge

valentinThomazic commented:
Hi @isa084 I am not sure about about the github CI but on the Gitlab (Thales) side your PR did not seem to add any issue.
There are test failing but this is not because of this PR.
